feat: add lance partitioning spec

wojiaodoubao opened the pull request:
This pr tries to add partition as an experimental spec. Discussion could be fount at: https://github.com/lance-format/lance-namespace/discussions/272

This PR remains a draft, and we still need to hold a vote to decide whether to introduce the partition spec.

jackye1995 commented on docs/src/rest/partition-spec.md:
```diff
@@ -0,0 +1,161 @@
+## Lance Partition Spec (Experimental)
```
One thought, can we make this a part of the DirectoryNamespace catalog spec, because it is basically a directory namespace + all tables must have the same schema, and the namespace must be leveled.

jackye1995 commented on docs/src/rest/partition-spec.md:
```diff
@@ -0,0 +1,161 @@
+This specification follows the "convention over configuration" principle, defining a partitioned table as a **metadata specification**. It is the responsibility of compatible clients or computation engines to interpret and enforce it.
+
+### 2. Core Concepts
+#### 2.1. Partitioned Table
```
I think calling this a partitioned table would be quite confusing to many people because the actual interface is namespace not table. What about "Partitioned Namespace"?

jackye1995 commented on docs/src/dir/partition-spec.md:
```diff
@@ -0,0 +1,141 @@
+#### 2.1. Metadata Definition
+A `Namespace` object is identified as the root of a partitioned namespace if its metadata properties contain the following key-value pairs:
+
+- `lance.partitioning.is_partitioned` (String): The value must be `"true"`. This explicitly identifies the Namespace as the root of a partitioned namespace.
```
this can just be `lance.partitioning.enabled`?

jackye1995 commented on docs/src/dir/partition-spec.md:
```diff
@@ -0,0 +1,141 @@
+## Lance Partition Spec (Experimental)
```
no need to say experimental. But I think the title should be "Lance Partitioned Namespace Spec"

jackye1995 commented on docs/src/dir/partition-spec.md:
```diff
@@ -0,0 +1,141 @@
+    - **Query Pruning**: Pruning can only be performed when the query condition provides the exact same value that was used for partitioning.
+
+### 4. Compatibility and Error Handling
+- **Backward Compatibility**: Clients that are not aware of this partitioning specification can still browse and operate on a partitioned namespace as a normal directory tree of namespaces and tables.
```
this seems unnecessary to state

jackye1995 commented on docs/src/dir/partition-spec.md:
```diff
@@ -0,0 +1,141 @@
+
+### 4. Compatibility and Error Handling
+- **Backward Compatibility**: Clients that are not aware of this partitioning specification can still browse and operate on a partitioned namespace as a normal directory tree of namespaces and tables.
+- **Schema Consistency**: When writing data, if the schema of the data to be written is incompatible with the defined schema of the partitioned namespace, an error should be returned. When creating a new leaf partition, the partitioned namespace's schema must be used.
```
we should expand this a bit more regarding the read and write behavior against the namespace. We should assume it is possible for users to operate directly against individual partition tables. Thus:
1. at read time, it should use the namespace schema. If table does not have the schema, it should return NULL as values, if table have extra columns, those columns are ignored. If there is unmatched data type, namespace schema data type should be used, if conversion not possible it should fail.
2. at write time, it should transform data to the table specific schema, following a similar rule as reader.

jackye1995 commented:
@wojiaodoubao I edited the doc with some additional designs, let me know if this looks good to you.

wojiaodoubao commented on docs/src/dir/partition-spec.md:
```diff
@@ -0,0 +1,390 @@
+
+### Namespace Naming
+
+Partition namespaces use **random identifier naming** to avoid issues with special characters in partition values. 
```
+1

wojiaodoubao commented on docs/src/dir/partition-spec.md:
```diff
@@ -0,0 +1,390 @@
+
+## Metadata Definition
+
+A directory namespace is identified as a partitioned namespace if the `__manifest` table's [metadata](catalog-spec.md#root-namespace-properties) contains at least one partition spec version key.
```
+1. The root path of the partitioned namespace will be the table location.

wojiaodoubao commented on docs/src/dir/partition-spec.md:
```diff
@@ -0,0 +1,390 @@
+The partitioning information is stored in `partition_spec_v<N>` (e.g., `partition_spec_v1`), which is a JSON array of partition field objects. Each partition field contains:
+
+* A **field id** uniquely identifying this partition field
+* A **name** for the partition field (used as the column name in `__manifest`)
```
Iceberg seems use `_partition` as the name of partition field. Do we use `name` to represent the partition field? What if the `name` conflicts with some column in schema?

jackye1995 replied:
> The Field ID represents the sequence. So if I have partition fields country, city, and I want to update it to continent , country, city, what should I do?

My thinking is that the sequence only matters because we need to know how to organize the namespace levels (as described in https://github.com/lance-format/lance-namespace/pull/279#discussion_r2659449766)

so if you add a `continent`, it will be `country -> city -> continent`. Technically for use cases like query pruning, partitioned join, I don't think this matters because you can still find "all partitions belong to continent Asia" without problem even though it is the last level not the first.

> First I need to resolve the field id issue: the filed id of country must be less than continent, so is city. I can do it by adding 3 new partition fields.

The partition specs are not independent, it should have some relationship between versions. In the example you give, we are adding a new partition field, so country and city field IDs should remain unchanged, and continent has the next field ID.

> Second I'll face name conflicts. Can I reuse country and city as partition field name?

that's a good point. I think the safest approach is likely to also use only id to reference the partition fields, so the column names would take `partition_field_{i}` in the `__manifest` table.

wojiaodoubao commented on docs/src/dir/partition-spec.md:
```diff
@@ -0,0 +1,390 @@
+
+1. Query engine analyzes the query predicate to identify filters on partition columns
+2. For each partition expression, the engine evaluates the expression with the query values to compute the expected partition value(s)
+3. Engine queries `__manifest` with filters on the partition columns
```
Will DirectoryNamespace providing a method to query `__manifest table`?

jackye1995 replied:
I think we already do, just need to make it `pub(crate)`: https://github.com/lance-format/lance/blob/main/rust/lance-namespace-impls/src/dir/manifest.rs#L535

jackye1995 commented:
Try to explain my thinking here:

1. versioned partition spec: partition evolution is a requirement to be comparable against similar projects, so we do need to version the partition spec

2. source field ID: lance table schema fields do have field ID, it's recorded in field metadata, so we should use that instead of field name to be schema evolution-friendly

3. partition field ID: partition field ID serves as the priority of partition keys, so we can still know the order of the fields even if it is converted to a map or something.

4. schema: the current approach is not fully correct yet, I still want to think a bit more on it. the schema id is used in spec as source id, so we need to know all fields that are still referenced in the partition specs, even if it is already dropped. Currently I just say we keep the unioned schema, but that might be not enough because the field ID could differ in namespace schema and underlying table schema. Maybe the easiest way is to just say in the spec that the namespace and table level schema should remain consistent.

5. partition expression: I switched to use just DataFusion expression to describe the partition transform, I feel this is more generic and also makes it more consistent with over Lance philosophy that we try to use Arrow & DataFusion semantics, rather than invent our own semantics for query related stuffs.

6. partition namespace name: if we use `col=val` style, if value has $ character it would not work. So we just use a random name and the partition key and value can be found in namespace properties (generated dynamically from __manifest column data). This also allow us to avoid leaking data information in the namespace name for users that want high security measures.

7. partition pruning: to make pruning work more efficiently, we directly store partition info as columns, and pruning is just a SQL call to select the table matching the predicate.

8. transaction support: I added the additional `read_version` column in __manifest so that if user desire multi-partition transaction, __manifest can be used as the main medium to coordinate commits and enforce read isolation. More details are probably needed in this section to describe the read and write path in more details.

wojiaodoubao commented:
> versioned partition spec: partition evolution is a requirement to be comparable against similar projects, so we do need to version the partition spec

+1


> source field ID: lance table schema fields do have field ID, it's recorded in field metadata, so we should use that instead of field name to be schema evolution-friendly

Does it mean all tables must be consistent at filed ID?


> partition field ID: partition field ID serves as the priority of partition keys, so we can still know the order of the fields even if it is converted to a map or something.

+1

> schema: the current approach is not fully correct yet, I still want to think a bit more on it. the schema id is used in spec as source id, so we need to know all fields that are still referenced in the partition specs, even if it is already dropped. Currently I just say we keep the unioned schema, but that might be not enough because the field ID could differ in namespace schema and underlying table schema. Maybe the easiest way is to just say in the spec that the namespace and table level schema should remain consistent.

+1 say in the spec that the namespace and table level schema should remain consistent.



> partition expression: I switched to use just DataFusion expression to describe the partition transform, I feel this is more generic and also makes it more consistent with over Lance philosophy that we try to use Arrow & DataFusion semantics, rather than invent our own semantics for query related stuffs.

+1. For engines that support DataFusion, we can directly use DataFusion to compute partition values; for those that do not, we need to implement functions with the same semantics ourselves, right?



> partition namespace name: if we use col=val style, if value has $ character it would not work. So we just use a random name and the partition key and value can be found in namespace properties (generated dynamically from __manifest column data). This also allow us to avoid leaking data information in the namespace name for users that want high security measures.

+1

> partition pruning: to make pruning work more efficiently, we directly store partition info as columns, and pruning is just a SQL call to select the table matching the predicate.

+1. I was thinking, how should we expose it to engine? It seems a bad idea to expose __manifest table directly. So maybe adding a new function to DirectoryNamespace(not Namespace) to perform partition pruning?


> transaction support: I added the additional read_version column in __manifest so that if user desire multi-partition transaction, __manifest can be used as the main medium to coordinate commits and enforce read isolation. More details are probably needed in this section to describe the read and write path in more details.

Current spec supports schema evolution, partition evolution, and ACID. Does this mean it adds an additional table format layer on top of the Lance Table Format? I wonder if we need to clarify which table format capabilities can be implemented at the partition layer and which will never be implemented there.

wojiaodoubao commented on docs/src/dir/partition-spec.md:
```diff
@@ -0,0 +1,390 @@
+
+**Field ID Stability**: Field IDs (`lance:field_id`) are never reused. Once a field ID is assigned, it permanently identifies that logical column even if the column is later deprecated. This ensures partition specs using `source_id` references remain valid.
+
+**Partition Field Validity**: If a source column is deprecated, existing partition fields referencing it via `source_id` remain valid for reading existing data. However, new partition spec versions should not reference deprecated columns. To remove a partition field, create a new partition spec version without that field.
```
> existing partition fields referencing it via `source_id` remain valid for reading existing data

In partitioned namespace we won't support time travel. After a filed is removed from schema, it seems we don't need the removed field when reading existing data. 

1. Engine analyzes query predicate and get partition expressions. The removed field doesn't exist in any predicate(otherwise it breaks semantic check since the removed field is not in schema), so it doesn't exist in any partition expression.
2. Engine evaluates the expression, compute the expected partition value(s), queries `__manifest` with filters, retrieves the paths of matching `dataset` tables, finally scan them.

So I think we don't need to mark source column as deprecated, instead maybe just remove it. We can also drop the `partition_field` column since there is noway to reference the partition field.

jackye1995 commented:
> For engines that support DataFusion, we can directly use DataFusion to compute partition values; for those that do not, we need to implement functions with the same semantics ourselves, right?

Yes. I'm not saying we are focusing engines to use DataFusion. But by stating the expression in it, we do not need to explain the behaviors of the expressions anymore, all the behaviors to all data types, null, computation behaviors are already defined. Any engine can make sure the behaviors match exactly to the DataFusion behavior when evaluating the expression. 

With that being said, I think most likely we will add some rust implementation to return a list of datasets to scan for a query, and then bind that to other languages, so most engines will eventually call into DataFusion to do the pruning.

jackye1995 commented:
> I was thinking capabilities like: branch, tag, time travel, deletion a physical partition(table), compaction, cleanup etc for partitioned namespace

I think there are 2 categories:
1. features in the table format: by the nature of using a __manifest Lance table, implementations of the partitioned namespace spec is able to do branching, time travel, deletion vector, etc. against contents in __manifest. I don't think we need to strictly enumerate all those features or ban any features, since the spec is mainly describing the storage layout and not all features it can not cannot do. 
2. features offered in rust SDK and engine implementations: for example compaction, cleanup, etc.  I think we will just start from the initial set of features like basic read and write support, and see whatever new features we need step by step. We might want to do some planning together. Although I put things like partition evolution in the spec, that's mainly to ensure the spec is future proof and we don't need to immediately do a v2 in short term. We don't have to implement it right away if features like compaction across partitions is more important.

wojiaodoubao commented:
> @wojiaodoubao I updated the entire doc based our discussion results. Let me know what you think!

Great work, Big +1!
